### PR TITLE
Allow sorting ascending/descending order for songs and albums

### PR DESCRIPTION
**Pull request by CappielloAntonio**
While I successfully (re)implemented the ability to sort ascending and descending, now the songs in album get ordered too. It works, maybe it's useful to somebody.

TODO: find a way to distinguish between caller. Any idea to do it elegantly? 😃

### CODE REVIEW

**Comment by dkanada** [comment]
Do you mean the finding the caller in the QueryUtil class?

**Comment by CappielloAntonio** [comment]
I have to distinguish if I call the get songs function in order to populate the songs fragment or in order to populate the album details fragment. 
I have an idea, I will check if it is feasible

**Comment by dkanada** [comment]
We might want to start using `TAG` for this. You can see some of the hacky checks I did in the album and artist methods to implement what you're thinking about at the moment.

**Comment by dkanada** [comment]
That actually doesn't sound ideal either. What about a flag system like a lot of AOSP components have?

**Comment by CappielloAntonio** [comment]
I may have found a solution, but I want to discuss it with you. 
We could change getSongs method signature to accept a bool flag which indicate if the list must be sorted or not, depending on the context.
This is the easiest solution, which make also sense, and theoretically could also be applied to the sort method properties in order to change this:

`if (query.getSortBy().length != 0) return;
`

But now the question: if I change the sort order method in SongsFragment, it means I want to change the order in GenreDetailActivity/ArtistDetailActivity? Is this setting a global valid setting or its power should be limited to the SongsFragment page?

**Comment by dkanada** [comment]
> Is this setting a global valid setting or its power should be limited to the SongsFragment page?

For now it should definitely be per-fragment like the sort method. I think you can actually see the individual settings in PreferenceUtils I had prepared for this feature.

**Comment by CappielloAntonio** [comment]
> For now it should definitely be per-fragment like the sort method. I think you can actually see the individual settings in PreferenceUtils I had prepared for this feature.

At the moment if I set a sort method (name/artist/added/...) in SongsFragment this will have effect everywhere I'll call QueryUtil's getSongs(), so GenreDetailActivity/ArtistDetailActivity, but not in AlbumDetailActivity. 

I've replicated the same behavior with the sort order methods (ascending/descending). If I set an order in SongsFragment I'll find my songs sorted GenreDetailActivity/ArtistDetailActivity but not in AlbumDetailActivity. 

Is it correct?

File: app/src/main/java/com/dkanada/gramophone/ui/fragments/mainactivity/library/pager/AbsLibraryPagerRecyclerViewCustomGridSizeFragment.java [hunk 1]
```diff
@@ -85,6 +94,13 @@ public void setAndSaveSortMethod(final String sortMethod) {
         invalidateAdapter();
     }
 
+    public void setAndSaveSortOrder(final String sortOrder) {
+        this.sortOrder = sortOrder;
+        saveSortOrder(sortOrder);
+        setSortOrder(sortOrder);
+        invalidateAdapter();
```
Comment (dkanada) [inline app/src/main/java/com/dkanada/gramophone/ui/fragments/mainactivity/library/pager/AbsLibraryPagerRecyclerViewCustomGridSizeFragment.java:101]: I just installed the new version and this doesn't seem to be refreshing the list for the sort method.

**Comment by dkanada** [comment]
I'm also wondering if we even need the query utilities file to be honest. Most of the methods have a small amount of uses, and each one usually requires one or two specific alterations to the query anyways.

File: app/src/main/java/com/dkanada/gramophone/ui/fragments/mainactivity/library/pager/AbsLibraryPagerRecyclerViewCustomGridSizeFragment.java (hunk 1 above)
Comment (CappielloAntonio) [reply app/src/main/java/com/dkanada/gramophone/ui/fragments/mainactivity/library/pager/AbsLibraryPagerRecyclerViewCustomGridSizeFragment.java:101]: Mmm, I just tried and it worked as expected. What did you tried exactly?

Comment (dkanada) [reply app/src/main/java/com/dkanada/gramophone/ui/fragments/mainactivity/library/pager/AbsLibraryPagerRecyclerViewCustomGridSizeFragment.java:101]: It was just taking longer than expected. When the method changes, it jumps to the top which looks like a refresh, and then actually replaces the data set a few seconds later. I'm currently making changes to the abstract classes to see if I can add a method for the queries there instead.

**Comment by dkanada** [comment]
I spent this weekend converting the custom ObservableRecyclerView elements into standard RecyclerView elements and AppBarLayouts in preparation for some design changes to the details activities. I think the best plan is to remove the song and album methods from the QueryUtils class. Those will almost always require custom modifications so it doesn't make much sense to try and standardize them. What do you think?

**Comment by CappielloAntonio** [comment]
> I spent this weekend converting the custom ObservableRecyclerView elements into standard RecyclerView elements and AppBarLayouts in preparation for some design changes to the details activities. I think the best plan is to remove the song and album methods from the QueryUtils class. Those will almost always require custom modifications so it doesn't make much sense to try and standardize them. What do you think?

I completely agree with you. Obviously I would have preferred to standardize and reuse the same methods everywhere, but unfortunately this is not the case.

**Comment by dkanada** [comment]
Do you want me to remove the album and song methods and then you could rebase this pull request, or did you want to handle it with the rebase?

**Comment by CappielloAntonio** [comment]
What behaviour are you trying to accomplish exactly? I now should be able to sort and reorder my lists correctly. 
If you want to remove the album and song methods does it means are you using more specific methods to sort them? 
Anyway, you can remove those two methods and I can rebase my pull request. I'll do what you think it's the best.
Always glad to help 😃

**Comment by dkanada** [comment]
I wanted to find the time to fix this properly, but I'd rather merge this sooner than later, so let's just find a good hack for now. I really don't like adding more parameters to the query methods, so how about we just remove the `applyProperties()` method for now and set those manually?

I remember the reason I created `QueryUtils` as well. If I ever add offline support, the app needs to fetch data from a remote server or just use a local cache. I figured the easiest way would be with a utility class, but perhaps I was wrong. I'll figure that out soon enough.

**Comment by dkanada** [comment]
By the way, I'm currently working on some changes for infinite scroll now that most of the annoying bugs have been fixed. Let me know if you're working on other things so we don't accidentally add the same feature :p

**Comment by dkanada** [comment]
Sorry it took so long to get this merged, I had to fix the conflicts from infinite scrolling.

**Pull request merged**